capstone/scripts/refactor_xml.py

```diff
@@ -0,0 +1,936 @@
+                number = re.match(r'\d+', headnote.text())
+                if number:
+                    headnotes_lookup[number.group(0)] = headnote.attr.id
+            for headnote_ref in casebody('bracketnum').items():
+                number = re.search(r'\d', headnote_ref.text())
+                if number and number.group(0) in headnotes_lookup:
+                    headnote_ref.attr.ref = headnotes_lookup[number.group(0)]
+
+        # handle each <opinion>
+        case['opinions'] = opinions = []
+        for footnote_opinion_index, (op_type, children) in enumerate(sections):
+
+            # link footnotes:
+            # add id='footnote_<opinion count>_<footnote count>' attribute to each <footnote>
+            # add ref='<footnote_id>' attribute to each <footnotemark>
+            if not duplicative:
+                footnote_index = 1
+                footnotes_lookup = {}
+                for tag in children:
+                    if tag[0].tag == 'footnote':
+                        footnote_id = 'footnote_%s_%s' % (footnote_opinion_index, footnote_index)
+                        tag.attr.id = footnote_id
+                        if tag.attr.label:
+                            footnotes_lookup[tag.attr.label] = footnote_id
+                        footnote_index += 1
+                for tag in children:
+                    for footnote_ref in tag.find('footnotemark').items():
+                        if footnote_ref.text() in footnotes_lookup:
+                            footnote_ref.attr.ref = footnotes_lookup[footnote_ref.text()]
+
+            # extract data for this opinion, and update alto data to match
+            opinion = {'type': op_type}
+            paragraphs, footnotes, paragraph_els = extract_paragraphs(children, case_id_to_alto_ids, blocks_by_id)
+            sync_alto_blocks_with_case_paragraphs(paragraph_els, blocks_by_id, case_id_to_alto_ids)
+            if paragraphs:
+                opinion['paragraphs'] = paragraphs
+            if footnotes:
+                opinion['footnotes'] = footnotes
+            opinions.append(opinion)
+
+        # handle <corrections>
+        if not duplicative:
+            corrections = list(case_el.children('correction').items())
+            if corrections:
+                case['corrections'] = [el.text() for el in corrections]
+
+        cases.append(case)
+
+    # encrypt redacted strings
+    if redacted_storage:
+        print("Encrypting pages")
+        for page in pages:
+            page_obj = PageStructure(blocks=page['blocks'])
+            page_obj.encrypt(key)
+            page['encrypted_strings'] = page_obj.encrypted_strings
+
+        # create an unencrypted copy for checking reversability
+        print("Decrypting pages")
+        decrypted_pages = deepcopy(pages)
+        for page in decrypted_pages:
+            page_obj = PageStructure(blocks=page['blocks'], encrypted_strings=page['encrypted_strings'])
+            page_obj.decrypt(key)
```

Comment:
Do we just decrypt to make sure it runs, or do we check the results in `page_obj` as well? I think I'm probably missing something.

Reply:
This decrypts the pages in decrypted_pages. Down below in `for page in decrypted_pages` we make sure we can reconstruct each ALTO <Page> element, which relies on the decryption succeeding.